### PR TITLE
feat: push GetUsers authorization filter to SQL

### PR DESCRIPTION
**Pull request by Emyrk**
Prereq to https://github.com/coder/coder/pull/8447

# What this does

This pushes the GetUsers authorization filter into SQL. So all authz checks are done by the SQL `WHERE` clause.

This is probably not a performance increase, but makes fetching users and the number of users correct if we change permissions.

In the current codebase we allow all users to read all other users. Because of this assumption, `GetUsers` was lazy in how it handled user counts. If permissions were changed, this value would be inaccurate and leak the total number of users to a caller who does not have permission to know this value.

Removes `GetAuthorizedUserCount` and `GetFilteredUserCount` as it is no longer needed.

### CODE REVIEW

File: coderd/database/dbauthz/dbauthz.go [hunk 1]
```diff
@@ -943,12 +935,10 @@ func (q *querier) GetFileTemplates(ctx context.Context, fileID uuid.UUID) ([]dat
...
 func (q *querier) GetFilteredUserCount(ctx context.Context, arg database.GetFilteredUserCountParams) (int64, error) {
-	prep, err := prepareSQLFilter(ctx, q.auth, rbac.ActionRead, rbac.ResourceUser.Type)
-	if err != nil {
-		return -1, xerrors.Errorf("(dev error) prepare sql filter: %w", err)
+	if err := q.authorizeContext(ctx, rbac.ActionRead, rbac.ResourceSystem); err != nil {
+		return -1, err
 	}
-	// TODO: This should be the only implementation.
-	return q.GetAuthorizedUserCount(ctx, arg, prep)
+	return q.db.GetFilteredUserCount(ctx, arg)
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/database/dbauthz/dbauthz.go]: Actually I could remove it

File: coderd/database/dbfake/dbfake.go [hunk 2]
```diff
@@ -5365,76 +5358,37 @@ func (q *FakeQuerier) GetAuthorizedWorkspaces(ctx context.Context, arg database.
...
-
-	if len(params.RbacRole) > 0 && !slice.Contains(params.RbacRole, rbac.RoleMember()) {
-		usersFilteredByRole := make([]database.User, 0, len(users))
-		for i, user := range users {
-			if slice.OverlapCompare(params.RbacRole, user.RBACRoles, strings.EqualFold) {
-				usersFilteredByRole = append(usersFilteredByRole, users[i])
-			}
-		}
-
-		users = usersFilteredByRole
```
Comment (Emyrk) [inline coderd/database/dbfake/dbfake.go:5436]: All this is done by the `GetUsers` call

File: coderd/database/dbauthz/dbauthz.go [hunk 3]
```diff
@@ -587,8 +587,8 @@ func (q *querier) SoftDeleteTemplateByID(ctx context.Context, id uuid.UUID) erro
 }
 
 func (q *querier) GetUsersWithCount(ctx context.Context, arg database.GetUsersParams) ([]database.User, int64, error) {
```
Comment (kylecarbs) [inline coderd/database/dbauthz/dbauthz.go:589]: Do we not need this function anymore? Can it just use `GetUsers` instead?

**Review by johnstcn — APPROVED** [review]
You can argue that this is a refactor

File: coderd/database/dbauthz/dbauthz.go (hunk 3 above)
Comment (Emyrk) [reply coderd/database/dbauthz/dbauthz.go:589]: Yup! Good catch, it was an extra function. I should write a unit test to error on extra funcs :thinking:

Comment (Emyrk) [reply coderd/database/dbauthz/dbauthz.go:589]: Ah I see, I did this thinking dbauthz wouldn't have to match the db.Store interface and just made the function signature nicer. I'll still remove it though

**Pull request merged**